libcst/metadata/scope_provider.py

```diff
@@ -0,0 +1,309 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+import abc
+import builtins
+from collections import defaultdict
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Dict, Iterator, List, MutableMapping, Optional, Tuple, Type, Union
+
+import libcst as cst
+from libcst.metadata.base_provider import BatchableMetadataProvider
+from libcst.metadata.expression_context_provider import (
+    ExpressionContext,
+    ExpressionContextProvider,
+)
+
+
+@dataclass(frozen=True)
+class Access:
+    node: cst.Name
+    scope: "Scope"
+
+
+class BaseAssignment:
+    name: str
+    scope: "Scope"
+    __accesses: List[Access]
+
+    def __init__(self, name: str, scope: "Scope") -> None:
+        self.name = name
+        self.scope = scope
+        self.__accesses = []
+
+    def record_access(self, access: Access) -> None:
+        self.__accesses.append(access)
+
+    @property
+    def accesses(self) -> Tuple[Access, ...]:
+        # we don't want to publicly expose the mutable version of this
+        return tuple(self.__accesses)
+
+
+class Assignment(BaseAssignment):
+    node: cst.CSTNode
+
+    def __init__(self, name: str, scope: "Scope", node: cst.CSTNode) -> None:
+        self.node = node
+        super().__init__(name, scope)
+
+
+class BuiltinAssignmemt(BaseAssignment):
+    node: None = None
```

Comment:
Since we don't actually have the CSTNode of the builtins' assignment. I intentionally made the node attribute not available in `BuiltinAssignmemt`. Will remove the `node=None`.